app/controllers/answers_controller.rb

```diff
@@ -0,0 +1,63 @@
+class AnswersController < ApplicationController
+  before_action :user_auth, except: %i[index]
+  before_action :set_answer, only: %i[update destroy]
+  before_action :validate_owner, only: %i[update destroy]
+  before_action :check_question_presence, only: %i[create update]
+
+  def index
+    render json: Answer.all, status: :ok
+  end
+
+  def create
+    answer = Answer.new(answer_params.merge(user_id: @current_user.id, question_id: answer_params[:question_id]))
+    return render json:  { message: answer.errors }, status: :unprocessable_entity unless answer.save
+
+    render json: answer, status: :created
+  end
+
+  def update
+    unless @answer.update(answer_params)
+      return render json: { message: @answer.errors.full_messages },
+                    status: :unprocessable_entity
+    end
+
+    render json: @answer, status: :ok
+  end
+
+  def destroy
+    unless @answer.destroy
+      return render json: { message: @answer.errors.full_messages },
+                    status: :unprocessable_entity
+    end
+
+    render json: @answer, status: :no_content
+  end
+
+  private
+
+  def check_question_presence
+    question = Question.find_by(id: answer_params[:question_id])
+    return render json: { message: 'question not found' }, status: :not_found if question.nil?
+  end
+
+  def validate_owner
+    return unless @answer.user_id != @current_user.id
+
+    message = case action_name
+              when 'update'
+                'Not authorized to update the answer'
+              when 'destroy'
+                'Not authorized to delete the answer'
+              end
+    render json: { message: }, status: :unauthorized
```

Comment:
return render json: {message: }, status: :unauthorized if @answer.user_id != @current_user.id

after that check is done and if the user_id is same as current_user's id then the execution will continue.

Reply:
what is wrong in this  one, and how can we return message if it is not set

Reply:
Your logic gives the right output, but when it comes to the readability of the code, it's very confusing. 

The problem with your approach is you're checking and returning if the @answer.user_id == @current_user.id at the beginning. What if you had to check more complicated condition. Then, It'd get more messed up.

So, the good practice is to check and handle errors at the beginning and then perform the operation if everything is fine.

You can follow this pseudo-code to rewrite your validate_owner middleware:

1. Set the message based on your action.
2. If @answer.user_id != @current_user.id then return render message
3. else the execution can continue